Image background

nam-van-nguyen opened the pull request:
Would you please review and merge the code that supports the following?
1. Themes with or without background images
2. Users can put additional .ini files in the config folder for additional theme
3. Users can edit colors of each theme to match their preference, with Reset button to restore to default values
4. Moved center UI area code from Widget to OpenGLWidget
5. Possibly improving some smoothness issues with scheduling changes and the upgrade to OpenGl widget

antartica committed:
Change background to white and adjust colors accordingly

nam-van-nguyen committed:
Add preference configurations for the play animation colors.
  + Added new fields in GUI preferences for various color fields
  + Moved the color settings over from CFG to CSettings, but keep the CFG methods still pointing to the CSettings due some dependency issue.
  + The color is 0-255 integer value for "red", "green" and "blue".
  + It would help if add a button for picking color, and even better if have a preview while changing the color.

nam-van-nguyen committed:
Added copyright statement to the new header file IColorPreference.h

nam-van-nguyen committed:
Updated to make the stave thickness
Reduced vertical bar thickness
Increased font size
Caching font object

nam-van-nguyen committed:
Added a color selector for each color setting so that user can click on the buttons to select a color

nam-van-nguyen committed:
Added an apply button to preference dialog so that users can view the change (while in listen mode) and make adjustment as needed.
  + The staves still don't update, needs to clear cache/redraw.

nam-van-nguyen committed:
Fixed first time loading not using default colors

nam-van-nguyen committed:
Fixed notes not shown when the note color changing is not set
Update background color when Apply button on the preference dialog is clicked
Update rendering of staves when Apply button is clicked on the preference dialog

nam-van-nguyen committed:
Added background color for the color selector buttons.  The button's background only changes when clicked applied.

nam-van-nguyen committed:
Fixed played code code not working due to previous change, put back the previous code.

nam-van-nguyen committed:
Remove menu color and menu selected color since they are not used.

nam-van-nguyen committed:
Removed all color SpinBoxes and uses the color picker only.

nam-van-nguyen committed:
Fix crashes due to calling CSettings::loadSettings at an earlier time

nam-van-nguyen committed:
Updated setting code to use cache for colors and also fix memory leak as the result

nam-van-nguyen committed:
Fixed application crash on exit due to deletion of a shared font object in Draw's destructor

nam-van-nguyen committed:
Added support for color themes
  + Added initial set of themes for color.
  + Additional theme files can be created and dropped into configuration folder
  + Each theme can be further customized to user's taste

nam-van-nguyen committed:
Minor change to Earth and Saturn themes

nam-van-nguyen committed:
Updates to various color themes

nam-van-nguyen committed:
Minor changes to Earth and Mars themes

nam-van-nguyen committed:
Added background image support
  + Added some sample images from https://publicdomainarchive.com/
  + Need to remove theme ini files in the user configuration directory to get latest changes
  + Only set the Default theme to have background color as an example

nam-van-nguyen committed:
Added support for alpha channel colors
Added more theme with image background
Added more images from https://www.rawpixel.com with public domain license
  - https://www.rawpixel.com/image/3109519/free-illustration-image-botanical-frame-watercolor-bird-red-design
  - https://www.rawpixel.com/image/3109526/free-illustration-image-vintage-animalia-volatilia-animal
  - https://images.rawpixel.com/image_1000/czNmcy1wcml2YXRlL3Jhd3BpeGVsX2ltYWdlcy93ZWJzaXRlX2NvbnRlbnQvbHIvcmVtaXhwcm9qZWN0NC1hZS0yMS1sLWpvYjU5OC5qcGc.jpg?s=R29veuJVAfoshuq7uQTetsd0pnoKNFcMx1dcQuhduJc

nam-van-nguyen committed:
Minor changes to the Galaxy theme's playing zone background color

nam-van-nguyen committed:
Improved code for copying resources to user's config. directory
Updated various theme colors
Added additional themes
Removed some un-used theme images

Martchus commented on src/GlView.h:
```diff
@@ -30,14 +30,15 @@
 #include <QTime>
 #include <QBasicTimer>
 #include <QGLWidget>
```
Is `QGLWidget` still required at all? (If not it would be nice to remove the include.)

nam-van-nguyen replied:
Thank you for the review Martchus.  I have removed the include QGLWidget statement.  I also added a required include QGLFormat, that somehow worked for me before.  To be sure, I cleaned up this time and rebuilt it.

Martchus reviewed (commented):
Lots of code changes. I suppose this needs also lots of testing.

Not sure whether it makes sense to add that many background images to Git.

nam-van-nguyen committed:
Removed un-used include of QtWidget in GlView
Added include QGLFormat in QtWindow

nam-van-nguyen committed:
Add 10 ticks to time to process a repaint

nam-van-nguyen commented:
FYI, I just added a new branch, https://github.com/nam-van-nguyen/PianoBooster/tree/keyboardcommands

Added support for using the Piano keyboard keys as command (88 keys):
               + Added preference keys startStopKeys (default "21,108"), rewindKeys (default "21,107"), speedUpKeys (default "21,105") and speedDownKeys (default "21,103")
               + These are comma separated list of integer Midi notes, when pressed in a same time, the application would start/stop, re-wind, speed up /spped down
               + For speed up/speed down, when press first time, it starts increase/decrease every half a second, press again to stop the speed change

The above keys combination are can be changed in the config. .ini file for as desired or depending on keyboard size.

Commit:
Updated code get it to run on MacOS Monterey
